File: src/main/scala/cn/academy/vanilla/teleporter/skill/ThreateningTeleport.scala
package cn.academy.vanilla.teleporter.skill

import cn.academy.ability.api.Skill
import cn.academy.ability.api.context.{ClientContext, ClientRuntime, Context, RegClientContext}
import cn.academy.core.client.sound.ACSounds
import cn.academy.vanilla.ModuleVanilla
import cn.academy.vanilla.teleporter.client.TPParticleFactory
import cn.academy.vanilla.teleporter.entity.EntityMarker
import cn.academy.vanilla.teleporter.util.TPSkillHelper
import cn.lambdalib2.s11n.network.NetworkMessage.Listener
import cn.lambdalib2.util._
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.{EnumHand, SoundCategory}
import net.minecraft.util.math.{RayTraceResult, Vec3d}

/**
  * @author WeAthFolD, KSkun
  */
object ThreateningTeleport extends Skill("threatening_teleport", 1) {

  @SideOnly(Side.CLIENT)
  override def activate(rt: ClientRuntime, keyID: Int) = activateSingleKey(rt, keyID, p => new TTContext(p))

}

object TTContext {

  final val MSG_EXECUTE = "execute"

}

import cn.lambdalib2.util.MathUtils._
import cn.academy.ability.api.AbilityAPIExt._
import TTContext._

class TTContext(p: EntityPlayer) extends Context(p, ThreateningTeleport) {

  private val exp: Float = ctx.getSkillExp
  private var attacked: Boolean = false

  @Listener(channel=MSG_KEYUP, side=Array(Side.CLIENT))
  private def l_onKeyUp() = {
    sendToServer(MSG_EXECUTE)
  }

  @Listener(channel=MSG_KEYABORT, side=Array(Side.CLIENT))
  private def l_onKeyAbort() = {
    terminate()
  }

  @Listener(channel=MSG_MADEALIVE, side=Array(Side.SERVER))
  private def s_madeAlive() = {
    if(player.getHeldItemMainhand == null) terminate()
  }

  @Listener(channel=MSG_TICK, side=Array(Side.SERVER))
  private def s_tick() = {
    if(player.getHeldItemMainhand == null) terminate()
  }

  @Listener(channel=MSG_EXECUTE, side=Array(Side.SERVER))
  private def s_execute() = {
    val curStack: ItemStack = player.getHeldItemMainhand
    if(curStack != null && ctx.consume(getOverload(exp), getConsumption(exp))) {
      attacked = true
      val result: TraceResult = calcDropPos
      var dropProb: Double = 1.0
      var attacked_ = false
      if(result.target != null) {
        attacked_ = true
        TPSkillHelper.attackIgnoreArmor(ctx, result.target, getDamage(curStack))
        ThreateningTeleport.triggerAchievement(player)
        dropProb = 0.3
      }
      if(!player.capabilities.isCreativeMode) {
        if({curStack.setCount(curStack.getCount-1); curStack.getCount} <= 0) player.setHeldItem(EnumHand.MAIN_HAND, null)
      }
      if(RandUtils.ranged(0, 1) < dropProb) {
        val drop: ItemStack = curStack.copy
        drop.setCount(drop.getCount-1)
        player.world.spawnEntity(new EntityItem(player.world, result.x, result.y, result.z, drop))
      }
      ctx.addSkillExp(getExpIncr(attacked_))
      ctx.setCooldown(lerpf(30, 15, exp).toInt)
    }

    sendToClient(MSG_EXECUTE, attacked.asInstanceOf[AnyRef])
    terminate()
  }

  private def getConsumption(exp: Float): Float = lerpf(35, 100, exp)

  private def getRange(exp: Float): Float = lerpf(8, 15, exp)

  private def getExpIncr(attacked: Boolean): Float = (if(attacked) 1 else 0.2f) * .003f

  private def getDamage(stack: ItemStack): Float = {
    var dmg: Float = lerpf(3, 6, ctx.getSkillExp)
    if(stack.getItem eq ModuleVanilla.needle) dmg *= 1.5f
    dmg
  }

  private def getOverload(exp: Float): Float = lerpf(18, 10, exp)

  def calcDropPos: TraceResult = {
    val range: Double = getRange(exp)
    var pos: RayTraceResult = Raytrace.traceLiving(player, range, EntitySelectors.living, BlockSelectors.filEverything)
    if(pos == null) pos = Raytrace.traceLiving(player, range, EntitySelectors.nothing)
    val ret: TraceResult = new TraceResult
    if(pos == null) {
      val mo = VecUtils.add(player.getPositionVector, multiply(player.getLookVec, range))
      ret.setPos(mo.x, mo.y, mo.z)
    }
    else if(pos.typeOfHit eq MovingObjectType.BLOCK) ret.setPos(pos.hitVec.x, pos.hitVec.y, pos.hitVec.z)
    else {
      val ent: Entity = pos.entityHit
      ret.setPos(ent.posX, ent.posY + ent.height, ent.posZ)
      ret.target = ent
    }
    ret
  }

}

@SideOnly(Side.CLIENT)
@RegClientContext(classOf[TTContext])
class TTContextC(par: TTContext) extends ClientContext(par) {

  private val COLOR_NORMAL: Color = new Color().fromHexColor(0xbabababa)
  private val COLOR_THREATENING: Color = new Color().fromHexColor(0xbab2232a)

  private var marker: EntityMarker = _

  @Listener(channel=MSG_TERMINATED, side=Array(Side.CLIENT))
  private def l_terminated() = {
    if(isLocal && marker != null) marker.setDead()
  }

  @Listener(channel=MSG_MADEALIVE, side=Array(Side.CLIENT))
  private def l_start() = {
    if(isLocal) {
      marker = new EntityMarker(player.world)
      player.world.spawnEntityInWorld(marker)
      marker.setPosition(player.posX, player.posY, player.posZ)
      marker.width = 0.5f
      marker.height = 0.5f
    }
  }

  @Listener(channel=MSG_TICK, side=Array(Side.CLIENT))
  private def l_tick() = {
    if(isLocal) {
      val res: TraceResult = par.calcDropPos
      if(res.target != null) res.y -= res.target.height
      marker.setPosition(res.x, res.y, res.z)
      marker.target = res.target
      marker.color = if(marker.target != null) COLOR_THREATENING else COLOR_NORMAL
    }
  }

  @Listener(channel=MSG_EXECUTE, side=Array(Side.CLIENT))
  private def c_end(attacked: Boolean) = {
    if(isLocal) marker.setDead()
    if(attacked) {
      ACSounds.playClient(player, "tp.tp", SoundCategory.AMBIENT, 0.5f)
      val dropPos: TraceResult = par.calcDropPos
      val dx: Double = dropPos.x + .5 - player.posX
      val dy: Double = dropPos.y + .5 - (player.posY - 0.5)
      val dz: Double = dropPos.z + .5 - player.posZ
      val dist: Double = MathUtils.length(dx, dy, dz)
      val mo: Motion3D = new Motion3D(player.posX, player.posY - 0.5, player.posZ, dx, dy, dz)
      mo.normalize
      var move: Double = 1
      var x: Double = move
      while(x <= dist) {
        {
          mo.move(move)
          player.world.spawnEntity(TPParticleFactory.instance.next(player.world, mo.getPosVec,
            new Vec3d(RandUtils.ranged(-.02, .02), RandUtils.ranged(-.02, .05), RandUtils.ranged(-.02, .02))))
        }
        move = RandUtils.ranged(1, 2)
        x += move
      }
    }
  }

}

class TraceResult {

  var x: Double = .0
  var y: Double = .0
  var z: Double = .0
  var target: Entity = _

  def setPos(_x: Double, _y: Double, _z: Double) {
    x = _x
    y = _y
    z = _z
  }

}

File: src/main/java/cn/academy/ability/vanilla/teleporter/skill/ThreateningTeleport.java
package cn.academy.ability.vanilla.teleporter.skill;

import cn.academy.ACItems;
import cn.academy.AcademyCraft;
import cn.academy.ability.Skill;
import cn.academy.ability.context.ClientContext;
import cn.academy.ability.context.ClientRuntime;
import cn.academy.ability.context.Context;
import cn.academy.ability.context.RegClientContext;
import cn.academy.ability.vanilla.teleporter.util.TPSkillHelper;
import cn.academy.advancements.ACAdvancements;
import cn.academy.client.render.misc.TPParticleFactory;
import cn.academy.client.sound.ACSounds;
import cn.academy.entity.EntityMarker;
import cn.lambdalib2.util.*;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.util.math.Vec3d;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import cn.lambdalib2.s11n.network.NetworkMessage.Listener;
import org.lwjgl.util.Color;

import static cn.lambdalib2.util.MathUtils.lerpf;

public class ThreateningTeleport extends Skill
{
    public static final ThreateningTeleport instance = new ThreateningTeleport();
    public ThreateningTeleport()
    {
        super("threatening_teleport", 1);
    }

    @SideOnly(Side.CLIENT)
    @Override
    public void activate(ClientRuntime rt, int keyID)
    {
        activateSingleKey2(rt, keyID, TTContext::new);
    }

    public static class TTContext extends Context
    {
        public static final String MSG_EXECUTE = "execute";
        private float exp = ctx.getSkillExp();
        private boolean attacked =false;

        public TTContext(EntityPlayer p)
        {
            super(p, ThreateningTeleport.instance);
        }

        @Listener(channel=MSG_KEYUP, side=Side.CLIENT)
        private void l_onKeyUp()
        {
            sendToServer(MSG_EXECUTE);
        }

        @Listener(channel=MSG_KEYABORT, side=Side.CLIENT)
        private void l_onKeyAbort()
        {
            terminate();
        }

        @Listener(channel=MSG_MADEALIVE, side=Side.SERVER)
        private void s_madeAlive()
        {
            if(player.getHeldItemMainhand().isEmpty()) terminate();
        }

        @Listener(channel=MSG_TICK, side=Side.SERVER)
        private void s_tick()
        {
            if(player.getHeldItemMainhand().isEmpty())
                terminate();
        }

        @Listener(channel=MSG_EXECUTE, side=Side.SERVER)
        private void s_execute()
        {
            ItemStack curStack = player.getHeldItemMainhand();
            if(!curStack.isEmpty() && ctx.consume(getOverload(exp), getConsumption(exp)))
            {
                attacked = true;
                TraceResult result = calcDropPos();
                double dropProb = 1.0;
                boolean attacked_ = false;
                if(result.target != null) {
                    attacked_ = true;
                    TPSkillHelper.attackIgnoreArmor(ctx, result.target, getDamage(curStack));
                    dropProb = 0.3;
                    }
                if(!player.capabilities.isCreativeMode) {
                    curStack.shrink(1);
                }
                if(RandUtils.ranged(0, 1) < dropProb) {
                    ItemStack drop = curStack.copy();
                    drop.setCount(1);
                    player.world.spawnEntity(new EntityItem(player.world, result.x, result.y, result.z, drop));
                }
//                if(exp>0.5) ACAdvancements.trigger(player, ACAdvancements.ac_milestone.ID);
                ctx.addSkillExp(getExpIncr(attacked_));
                ctx.setCooldown((int)lerpf(30, 15, exp));
            }
            sendToClient(MSG_EXECUTE, attacked);
            terminate();
        }

        private float getConsumption(float exp)
        {
            return lerpf(35, 100, exp);
        }

        private float getRange(float exp)
        {
            return lerpf(8, 15, exp);
        }

        private float getExpIncr(boolean attacked)
        {
            return (attacked?1:0.2f)*0.003f;
        }

        private float getDamage(ItemStack stack)
        {
            float dmg = lerpf(3, 6, ctx.getSkillExp());
            if(stack.getItem()== ACItems.needle) dmg *= 1.5f;
            return dmg;
        }

        private float getOverload(float exp)
        {
            return lerpf(18, 10, exp);
        }

        public TraceResult calcDropPos()
        {
            double range = getRange(exp);
            RayTraceResult pos = Raytrace.traceLiving(player, range, EntitySelectors.living(), BlockSelectors.filEverything);
            if(pos.typeOfHit == RayTraceResult.Type.MISS)
                pos = Raytrace.traceLiving(player, range, EntitySelectors.nothing());
            TraceResult ret = new TraceResult();
            if(pos.typeOfHit == RayTraceResult.Type.MISS) {
                Vec3d mo = VecUtils.add(player.getPositionEyes(1.0F), VecUtils.multiply(player.getLookVec(), range));
                ret.setPos(mo.x, mo.y, mo.z);
            }
            else if(pos.typeOfHit == RayTraceResult.Type.BLOCK) ret.setPos(pos.hitVec.x, pos.hitVec.y, pos.hitVec.z);
            else {
                Entity ent = pos.entityHit;
                ret.setPos(ent.posX, ent.posY + ent.height, ent.posZ);
                ret.target = ent;
            }
            return ret;
        }
    }

    @SideOnly(Side.CLIENT)
    @RegClientContext(TTContext.class)
    public static class TTContextC extends ClientContext
    {
        private Color COLOR_NORMAL = new Color(0xba,0xba,0xba,0xba);
        private Color COLOR_THREATENING = new Color(0xba,0xb2,0x23,0x2a);
        private EntityMarker marker = null;
        private TTContext par;
        private static final String MSG_EXECUTE = TTContext.MSG_EXECUTE;

        public TTContextC(TTContext par)
        {
            super(par);
            this.par = par;
        }

        @Listener(channel=MSG_TERMINATED, side=Side.CLIENT)
        public void l_terminated()
        {
            if(isLocal() && marker != null) marker.setDead();
        }

        @Listener(channel=MSG_MADEALIVE, side=Side.CLIENT)
        public void l_start(){
            if(isLocal()) {
                marker = new EntityMarker(player.world);
                player.world.spawnEntity(marker);
                marker.setPosition(player.posX, player.posY, player.posZ);
                marker.width = 0.5f;
                marker.height = 0.5f;
            }
        }

        @Listener(channel=MSG_TICK, side=Side.CLIENT)
        public void l_tick()
        {
            if(isLocal()) {
                TraceResult res = par.calcDropPos();
                if(res.target != null) res.y -= res.target.height;
                marker.setPosition(res.x, res.y, res.z);
                marker.target = res.target;
                marker.color = (marker.target != null)? COLOR_THREATENING : COLOR_NORMAL;
            }
        }

        @Listener(channel=MSG_EXECUTE, side=Side.CLIENT)
        public void c_end(boolean attacked)
        {
            if(isLocal()) marker.setDead();
            if(attacked)
            {
                ACSounds.playClient(player, "tp.tp", SoundCategory.AMBIENT, 0.5f);
                TraceResult dropPos = par.calcDropPos();
                double dx = dropPos.x + .5 - player.posX;
                double dy = dropPos.y + .5 - (player.posY - 0.5);
                double dz = dropPos.z + .5 - player.posZ;
                double dist = MathUtils.length(dx, dy, dz);
                double posX = player.posX;
                double posY = player.posY - 0.5;
                double posZ = player.posZ;
                Vec3d lookingVec = new Vec3d(dx, dy, dz).normalize();

                double move = 1;
                double x = move;
                while(x <= dist)
                {
                    posX += move*lookingVec.x;
                    posY += move*lookingVec.y;
                    posZ += move*lookingVec.z;
                    player.world.spawnEntity(TPParticleFactory.instance.next(player.world, new Vec3d(posX, posY, posZ),
                            new Vec3d(RandUtils.ranged(-.02, .02), RandUtils.ranged(-.02, .05), RandUtils.ranged(-.02, .02))));
                    move = RandUtils.ranged(1, 2);
                    x += move;
                }
            }
        }
    }

    public static class TraceResult {
        public double x = .0;
        public double y = .0;
        public double z = .0;
        public Entity target = null;

        public void setPos(double _x, double _y, double _z)
        {
            x = _x;
            y = _y;
            z = _z;
        }

    }
}
